src/coreclr/src/vm/processdiagnosticsprotocolhelper.cpp

```diff
@@ -0,0 +1,210 @@
+        return false;
+
+    memcpy(bufferCursor, &stringLen, sizeof(stringLen));
+    bufferCursor += sizeof(stringLen);
+
+    memcpy(bufferCursor, value, stringLen * sizeof(T));
+    bufferCursor += stringLen * sizeof(T);
+
+    bufferLen -= totalStringSizeInBytes.Value();
+    return true;
+}
+
+uint16_t ProcessInfoPayload::GetSize()
+{
+    LIMITED_METHOD_CONTRACT;
+
+    // The protocol buffer is defined as:
+    // X, Y, Z means encode bytes for X followed by bytes for Y followed by bytes for Z
+    // uint = 4 little endian bytes
+    // long = 8 little endian bytes
+    // GUID = 16 little endian bytes
+    // wchar = 2 little endian bytes, UTF16 encoding
+    // array<T> = uint length, length # of Ts
+    // string = (array<wchar> where the last char must = 0) or (length = 0)
+
+    // uint64_t ProcessId;
+    // GUID RuntimeCookie;
+    // LPCWSTR CommandLine;
+    // LPCWSTR OS;
+    // LPCWSTR Arch;
+
+    S_UINT16 size = S_UINT16(0);
+    size += sizeof(ProcessId);
+    size += sizeof(RuntimeCookie);
+
+    size += sizeof(uint32_t);
+    size += CommandLine != nullptr ?
+        S_UINT16(GetStringLength(CommandLine) * sizeof(WCHAR)) :
+        S_UINT16(0);
+
+    size += sizeof(uint32_t);
+    size += OS != nullptr ?
+        S_UINT16(GetStringLength(OS) * sizeof(WCHAR)) :
+        S_UINT16(0);
+
+    size += sizeof(uint32_t);
+    size += Arch != nullptr ?
+        S_UINT16(GetStringLength(Arch) * sizeof(WCHAR)) :
+        S_UINT16(0);
+
+    ASSERT(!size.IsOverflow());
+    return size.Value();
+}
+
+bool ProcessInfoPayload::Flatten(BYTE * &lpBuffer, uint16_t &cbSize)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+        PRECONDITION(cbSize == GetSize());
+        PRECONDITION(lpBuffer != nullptr);
+    }
+    CONTRACTL_END;
+
+    bool fSuccess = true;
+    // uint64_t ProcessId;
+    memcpy(lpBuffer, &ProcessId, sizeof(ProcessId));
+    lpBuffer += sizeof(ProcessId);
+    cbSize -= sizeof(ProcessId);
+
+    // GUID RuntimeCookie;
+    memcpy(lpBuffer, &RuntimeCookie, sizeof(RuntimeCookie));
+    lpBuffer += sizeof(RuntimeCookie);
+    cbSize -= sizeof(RuntimeCookie);
+
+    // LPCWSTR CommandLine;
+    fSuccess &= TryWriteString(lpBuffer, cbSize, CommandLine);
+
+    // LPCWSTR OS;
+    if (fSuccess)
+        fSuccess &= TryWriteString(lpBuffer, cbSize, OS);
+
+    // LPCWSTR Arch;
+    if (fSuccess)
+        fSuccess &= TryWriteString(lpBuffer, cbSize, Arch);
+
+    // Assert we've used the whole buffer we were given
+    ASSERT(cbSize == 0);
+
+    return fSuccess;
+}
+
+void ProcessDiagnosticsProtocolHelper::GetProcessInfo(DiagnosticsIpc::IpcMessage& message, IpcStream *pStream)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+        PRECONDITION(pStream != nullptr);
+    }
+    CONTRACTL_END;
+
+    struct ProcessInfoPayload payload = {};
+
+    // Get cmdline
+    payload.CommandLine = GetManagedCommandLine();
```

Comment:
If the path has a single or double quote in it, is it escaped? And then if it has a space in it, is the executable portion of the path surrounded by double quotes?

For example, if the path was

`\some\pa"th\wi th\fun\characters`

and arguments 

`arg1` `arg2` `arg ument3`

does it come back as

`"\some\pa\"th\wi th\fun\characters" arg1 arg2 "arg ument3"`

? You might have to do some escaping or manipulation here.

Reply:
My intent was to specifically tailor this logic to the CI test environment, as I'm not sure it's worth the time investment to write a super-robust commandline parser for a single test.  That being said, if there were an existing parser that could handle all these cases I would prefer to use that.  It might be overkill, but I guess I could use the System.CommandLine front-end parser for that purpose.

Another alternative is using pinvokes into native code on each platform to retrieve the same info that runtime is getting, e.g., read `/proc/pid/cmdline` on Linux, and not do any normalization at all.  Then I'm not generating a hypothetical commandline to compare against.  It just means writing a bunch of platform dependent pinvokes for a single test.

Reply:
I hope we don't need to parse or re-format the command line strings within the runtime, but certainly we should document what format clients can expect to see, or document that it matches the output of a particular OS API call.

"Tailoring to the CI environment" sounds like it was meant to respond to a comment below about the test code? I assume the product code is intended to work for all apps, not only the ones that run in .NET's CI environment : )

Reply:
Any normalization in the PR is happening strictly in the test logic.  This code doesn't manipulate the commandline that the runtime retrieves.  The runtime's value is generated in `SaveManagedCommandLine` @ `ceeload.cpp:11343`.  That calls `GetCommandLineW` on Windows while the commandline is recreated from `argv` on non-Windows in `Initialize` @ `pal.cpp:311`.  The non-Widows logic gets the full path of the executable and then appends the arguments from `argv[1..]` with spaces between each entry.

Reply:
I went back to the simple normalization to be more robust.  Trying to be too targeted resulted in the test being more brittle.  For instance, when tests are run on Windows, it looks like the script has 2 spaces between `corerun` and the first argument.  The simplified version would need to check for 2 spaces specifically which would break if someone ran the test manually and had any other number of spaces.  I tried using `Environment.CommandLine`, but the native code that collects that information does a fair amount normalizing/formatting itself.